js/flexbuffers.js

```diff
@@ -0,0 +1,1015 @@
+    const byteWidth = flexbuffers.BitWidth.toByteWidth(width);
+    offset += flexbuffers.BitWidth.paddingSize(offset, byteWidth);
+    return byteWidth;
+  }
+
+  function computeOffset(newValueSize) {
+    const targetOffset = offset + newValueSize;
+    let size = buffer.byteLength;
+    const prevSize = size;
+    while (size < targetOffset) {
+      size <<= 1;
+    }
+    if (prevSize < size) {
+      const prevBuffer = buffer;
+      buffer = new ArrayBuffer(size);
+      view = new DataView(buffer);
+      new Uint8Array(buffer).set(new Uint8Array(prevBuffer), 0);
+    }
+    return targetOffset;
+  }
+
+  function pushInt(value, width) {
+    if (width === flexbuffers.BitWidth.WIDTH8) {
+      view.setInt8(offset, value);
+    } else if (width === flexbuffers.BitWidth.WIDTH16) {
+      view.setInt16(offset, value, true);
+    } else if (width === flexbuffers.BitWidth.WIDTH32) {
+      view.setInt32(offset, value, true);
+    } else if (width === flexbuffers.BitWidth.WIDTH64) {
+      view.setBigInt64(offset, BigInt(value), true);
+    } else {
+      throw `Unexpected width: ${width} for value: ${value}`;
+    }
+  }
+
+  function pushUInt(value, width) {
+    if (width === flexbuffers.BitWidth.WIDTH8) {
+      view.setUint8(offset, value);
+    } else if (width === flexbuffers.BitWidth.WIDTH16) {
+      view.setUint16(offset, value, true);
+    } else if (width === flexbuffers.BitWidth.WIDTH32) {
+      view.setUint32(offset, value, true);
+    } else if (width === flexbuffers.BitWidth.WIDTH64) {
+      view.setBigUint64(offset, BigInt(value), true);
+    } else {
+      throw `Unexpected width: ${width} for value: ${value}`;
+    }
+  }
+
+  function writeInt(value, byteWidth) {
+    const newOffset = computeOffset(byteWidth);
+    pushInt(value, flexbuffers.BitWidth.fromByteWidth(byteWidth));
+    offset = newOffset;
+  }
+
+  function writeBlob(arrayBuffer) {
+    const length = arrayBuffer.byteLength;
+    const bitWidth = flexbuffers.BitWidth.width(length);
+    const byteWidth = align(bitWidth);
+    writeInt(length, byteWidth);
+    const blobOffset = offset;
+    const newOffset = computeOffset(length);
+    new Uint8Array(buffer).set(new Uint8Array(arrayBuffer), blobOffset);
+    stack.push(offsetStackValue(blobOffset, flexbuffers.ValueType.BLOB, bitWidth));
+    offset = newOffset;
+  }
+
+  function writeString(str) {
+    if (stringCache.hasOwnProperty(str)) {
+      stack.push(stringCache[str]);
+      return;
+    }
+    const utf8 = toUTF8Array(str);
+    const length = utf8.length;
+    const bitWidth = flexbuffers.BitWidth.width(length);
+    const byteWidth = align(bitWidth);
+    writeInt(length, byteWidth);
+    const stringOffset = offset;
+    const newOffset = computeOffset(length + 1);
+    new Uint8Array(buffer).set(utf8, stringOffset);
+    const stackValue = offsetStackValue(stringOffset, flexbuffers.ValueType.STRING, bitWidth);
+    stack.push(stackValue);
+    stringCache[str] = stackValue;
```

Comment:
this appears to always use the string cache, making constructing a new buffer cause unnecessary many allocations for a use case where strings don't repeat (much). This should be entirely optional. See C++ implementation.

Reply:
Not only that, it also means that in case of lots of arbitrary strings memory will just keep growing. It's not really cache-like behaviour.

Reply:
I renamed the caches to lookups. As those are actually lookup tables in order to perform deduplication. The deduplication strategies for strings, keys and vector of keys are enabled by default, but user can disable them when they create a builder or when they encode an object. So this way if performance is more important then payload size, they can be disabled.